### PR TITLE
Improve "merkledag: not found" error

### PR DESCRIPTION
**Pull request by hsanjuan**
Based on #33, taking advantage of the rather new Is() methods from the errors
package to check for equality.

### CODE REVIEW

File: merkledag.go [hunk 1]
```diff
@@ -7,7 +7,37 @@ import (
 	cid "github.com/ipfs/go-cid"
 )
 
-var ErrNotFound = fmt.Errorf("merkledag: not found")
+// ErrNotFound is used to signal when a Node could not be found. The specific
+// meaning will depend on the DAGService implementation, which may be trying
+// to read nodes locally but also, trying to find them remotely.
+var ErrNotFound = ErrNotFoundCid{}
```
Comment (Stebalien) [inline merkledag.go]: Returning `ErrNotFoundCid{Cid}` is already going to break anything comparing against `ErrNotFound`. Given that, I'd rather _explicitly_ break it and replace `ErrNotFound` with `type ErrNotFound struct { Cid }`.

**Comment by Stebalien** [comment]
We should consider the comments form #33 and implement a `NotFound` method on the error. See: https://github.com/ipfs/go-datastore/blob/9e9a3a8350fe9b372e4d1cf8dc590d1f005a1540/datastore.go#L202-L213.

**Comment by Stebalien** [comment]
Note: we're definitely going to punt this till after 0.5 as it may have other consequences.

File: merkledag.go (hunk 1 above)
Comment (hsanjuan) [reply merkledag.go]: Ok, but no-one returns ErrNotFoundCid (and we can grep-fix all of our things accordingly). So for everyone else things would still work. But yeah, let's play safe and break.

**Comment by hsanjuan** [comment]
> We should consider the comments form #33 and implement a `NotFound` method on the error. See: https://github.com/ipfs/go-datastore/blob/9e9a3a8350fe9b372e4d1cf8dc590d1f005a1540/datastore.go#L202-L213.

What is the rationale for this approach? I don't fully understand it. Is the rationale here to be able to check if an error implements a `NotFound` interface and then use that method to decide? Who is supposed to call a .NotFound method?

I checked the blog post and my take is that wrapping and letting the original packages decide if it's one of their `NotFound` is the accepted approach. Particularly now that Go added wrap() and Is() methods natively. This allows both treating them as opaque and checking them when needed. It also allows every package to have their own ErrNotFound (which can wrap other ErrNotFounds).

Am I missing something?

File: merkledag.go [hunk 2]
```diff
@@ -7,7 +7,37 @@ import (
...
+
+// Is allows to check whether any error is of this ErrNotFoundCid type.
+// Do not use this directly, but rather errors.Is(yourError, ErrNotFound).
+func (e ErrNotFoundCid) Is(err error) bool {
+	switch err.(type) {
+	case ErrNotFoundCid:
+		return true
+	default:
+		return false
+	}
```
Comment (hsanjuan) [inline merkledag.go]: We could actually do:

```
	var f interface{
		NotFound() bool
	}
	return errors.As(err, &f)
```
This might be better as it would catch any target that implements a NotFound(). Wdyt?

File: merkledag.go [hunk 3]
```diff
@@ -7,7 +7,39 @@ import (
...
+
+// Is allows to check whether any error is of this ErrNotFound type.
+// Do not use this directly, but rather errors.Is(yourError, ErrNotFound).
+func (e ErrNotFound) Is(err error) bool {
+	switch err.(type) {
+	case ErrNotFound:
+		return true
+	default:
+		return false
+	}
```
Comment (hsanjuan) [inline merkledag.go]: We could actually do:

```
	var f interface{
		NotFound() bool
	}
	return errors.As(err, &f)
```

This might then catch any target that implements NotFound, regardless of its actual type. So a user might call `errors.Is(ipldErrNotFound, datastoreErrNotFound)` and get  true. I am not sure we want this though, wdty?

File: merkledag.go [hunk 4]
```diff
@@ -7,7 +7,39 @@ import (
...
 
-var ErrNotFound = fmt.Errorf("merkledag: not found")
+// ErrNotFound is used to signal when a Node could not be found. The specific
+// meaning will depend on the DAGService implementation, which may be trying
+// to read nodes locally but also, trying to find them remotely.
+//
+// The Cid field can be filled in to provide additional context.
+type ErrNotFound struct {
+	Cid cid.Cid
+}
```
Comment (hsanjuan) [inline merkledag.go]: Exposed Cid thinking that this might be used from outside this module.

It is ugly though that you have to instantiate the struct whenever you want to check if `errors.Is(whatever, ErrNotFound{})`.

**Comment by Stebalien** [comment]
From async discussion:

Really, there are two kinds of checks: "is" and "property".

* is: Useful when you want to check if an error was ultimately caused by some specific error. Honestly, that's probably what we want here.
* property: This is how, e.g., timeout errors usually work. You implement functions on your custom error type indicating the properties your error has. Then, you can call `As(err, &targetInterface)` on the error to tell if the error _can_ have the given property. Finally, you invoke the actual function (in this case, `NotFound() bool` to determine if the error actually _has_ that property.

TL;DR: `NotFound()` is probably a "nice to have". For our particular use-case, it's not really all that useful as we specifically want to know if the error was "IPLD node not found", not some random "datastore is broken because the index wasn't found" error.

File: merkledag.go (hunk 4 above)
Comment (Stebalien) [reply merkledag.go]: We _can_ add an `IsNotFound` function.

File: merkledag.go (hunk 3 above)
Comment (Stebalien) [reply merkledag.go]: Thinking about this, we probably don't want to do this. "Random file not found" is not the same as "dag node not found".

Comment (hsanjuan) [reply merkledag.go]: yeah, it would be mixing the As() an Is() approaches.

File: merkledag.go (hunk 4 above)
Comment (hsanjuan) [reply merkledag.go]: Added, and made it equivalent to `Is`.
